python/ray/ml/examples/pytorch/permuted_mnist_domain_incremental_learning_example.py

```diff
@@ -0,0 +1,297 @@
+class SimpleMLP(nn.Module):
+    def __init__(self, num_classes=10, input_size=28 * 28):
+        super(SimpleMLP, self).__init__()
+
+        self.features = nn.Sequential(
+            nn.Linear(input_size, 512),
+            nn.ReLU(inplace=True),
+            nn.Dropout(),
+        )
+        self.classifier = nn.Linear(512, num_classes)
+        self._input_size = input_size
+
+    def forward(self, x):
+        x = x.contiguous()
+        x = x.view(-1, self._input_size)
+        x = self.features(x)
+        x = self.classifier(x)
+        return x
+
+
+def get_mnist_dataset(train: bool = True) -> ray.data.Dataset:
+    """Returns MNIST Dataset as a ray.data.Dataset.
+
+    Args:
+        train: Whether to return the train dataset or test dataset.
+    """
+
+    def mnist_dataset_factory():
+        torchvision.datasets.MNIST("./data", download=True, train=train)
+
+    mnist_dataset = ray.data.read_datasource(
+        SimpleTorchDatasource(), dataset_factory=mnist_dataset_factory
+    )
+
+    # TODO: provide a better way to apply torchvision transforms.
+    # Do some initial preprocessing on the PIL Image directly, then convert to numpy.
+    if train:
+        mnist_dataset = mnist_dataset.map_batches(
+            lambda batch: [
+                (RandomCrop(28, padding=4)(image), label) for (image, label) in batch
+            ]
+        )
+    mnist_dataset = mnist_dataset.map_batches(
+        lambda batch: [(np.asarray(image), label) for (image, label) in batch]
+    )
+
+    return mnist_dataset
+
+
+class PermutedMNISTStream:
+    """Generates streams of permuted MNIST Datasets.
+
+    Args:
+        n_tasks: The number of tasks to generate.
+
+    """
+
+    def __init__(self, n_tasks: int = 3):
+        self.n_tasks = n_tasks
+        self.permutations = [
+            np.random.permutation(28 * 28) for _ in range(self.n_tasks)
+        ]
+
+    def random_permute_dataset(
+        self, dataset: ray.data.Dataset, permutation: np.ndarray
+    ):
+        """Randomly permutes the pixels for each image in the dataset."""
+
+        # permutation transformation
+        class PixelsPermutation(object):
+            def __call__(self, batch):
+                permuted_batch = [
+                    (image.reshape(-1)[permutation].reshape(28, 28), label)
+                    for (image, label) in batch
+                ]
+                return permuted_batch
+
+        return dataset.map_batches(PixelsPermutation, compute=ActorPoolStrategy())
+
+    def generate_train_stream(self) -> Iterator[ray.data.Dataset]:
+        for permutation in self.permutations:
+            mnist_dataset = get_mnist_dataset(train=True)
+            permuted_mnist_dataset = self.random_permute_dataset(
+                mnist_dataset, permutation
+            )
+            yield convert_dataset_to_pandas_format(permuted_mnist_dataset)
+
+    def generate_test_stream(self) -> Iterator[ray.data.Dataset]:
+        for permutation in self.permutations:
```

Comment:
ha, so the training and testing datasets are based off the same set of permutations? :)
it's fine as an example, just curious.

Reply:
Yeah, I think this is expected for all cases. In general, for incremental learning, the model can learn multiple tasks, but is not expected to generalize beyond the tasks that it has learned on.